infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java

```diff
@@ -108,16 +118,50 @@ public StreamLogFiles(ServerContext serverContext, boolean noVerify) {
         verifyLogs();
         // Starting address initialization should happen before
         // initializing the tail segment (i.e. initializeMaxGlobalAddress)
-        initializeStartingAddress();
-        initializeMaxGlobalAddress();
+        startingAddress = serverContext.getStartingAddress();
+        long firstSegment = startingAddress / RECORDS_PER_LOG_FILE;
+        lastSegment = serverContext.getTailSegment();
+        logMetadata = initializeLogMetadata(firstSegment, lastSegment);
 
         // This can happen if a prefix trim happens on
         // addresses that haven't been written
-        if (Math.max(getGlobalTail(), 0L) < getTrimMark()) {
+        if (Math.max(logMetadata.getGlobalTail(), 0L) < getTrimMark()) {
             syncTailSegment(getTrimMark() - 1);
         }
     }
 
+    /**
+     * This method will scan the log (i.e. read all log segment files)
+     * on this LU and create a map of stream offsets and the global
+     * addresses seen.
+     * @param startSegment first segment in the log to be scanned
+     * @param endSegment the last segment to be scanned
+     * @return LogMetadata constructed from all the addresses in the
+     * consecutive segments from [startSegment, endSegment]
+     */
+    private LogMetadata initializeLogMetadata(long startSegment, long endSegment) {
+        LogMetadata metadata = new LogMetadata();
+        long start = System.currentTimeMillis();
+        for (long currentSegment = startSegment; currentSegment <= endSegment; currentSegment++) {
+            // TODO(Maithem): factor out getSegmentHandleForAddress to allow getting
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1135)